### ISSUE TITLE
DNS alias mode

### ISSUE THREAD

**Issue by brecht-vermeersch** [vancluever/terraform-provider-acme#358]
Has anything changed since the original issue?

https://github.com/vancluever/terraform-provider-acme/issues/196

**Comment by vancluever** [comment]
@brecht-vermeersch looks like lego now includes CNAME support automatically, which means we would too! This was added in [v4.9.0](https://github.com/go-acme/lego/blob/master/CHANGELOG.md#v490---2022-10-03).

Thanks!

**Comment by brecht-vermeersch** [comment]
@vancluever Hi, thanks for the quick response. Any examples on how you can configure this in the terraforms?

**Comment by brecht-vermeersch** [comment]
nvm it just worked ;)

**Comment by vancluever** [comment]
@brecht-vermeersch yep! It should be completely automatic without intervention. The only real detail you'd need to be cognizant of is the alias -> target setup. I might clarify that a bit more over what I added in 4610f7d. In the meantime, glad it works!

**Comment by yaroslav-nakonechnikov** [comment]
@brecht-vermeersch @vancluever but still it would be helpful to add some info how to use dns alias. 

for example, for aws it asks zone host id. and which it should be? original? if yes, it will be from another account and there is not allowed to do any changes.

**Comment by vancluever** [comment]
@yaroslav-nakonechnikov this will vary from provider to provider. WRT route53, the hosted zone ID is only needed to override any auto-detection, so normally it should not be needed. For the basic CNAME instructions you can check out the Let's Encrypt page at https://letsencrypt.org/2019/10/09/onboarding-your-customers-with-lets-encrypt-and-acme/#the-advantages-of-a-cname.

in the article's example, `_acme-challenge.www.example.com` is aliased to `www.example.com.validationserver.example.net`, and if this was hosted on route53, we will automatically follow the CNAME and provision the record at that target without any additional configuration, not even the hosted zone ID, so long as the provider is able to query for the domain with the configured credentials.

**Comment by yaroslav-nakonechnikov** [comment]
@vancluever  thank you! 
i'm checking it now... but previously it wasn't working, because we have 2 zones: public and private. And zones are have same domain name. 
So i had to override it.

**Comment by vancluever** [comment]
@yaroslav-nakonechnikov the split-horizon scenario will definitely be the trickier one as it may mess with CNAME resolution. You might want to look at [`recursive_nameservers`](https://registry.terraform.io/providers/vancluever/acme/latest/docs/resources/certificate#recursive_nameservers-1) and [`disable_complete_propagation`](https://registry.terraform.io/providers/vancluever/acme/latest/docs/resources/certificate#disable_complete_propagation-1) too.

Remember that you will likely need to provision everything on the external zone as Let's Encrypt (or your other external ACME CA) will need to query the nameserver for the challenge records too.

Hope this helps!

**Comment by yaroslav-nakonechnikov** [comment]
so, yes,  i've tried it, but it fails with:
```
Error: error creating certificate: error: one or more domains had a problem:
[*.dev.some.nice.domain] [*.dev.some.nice.domain] acme: error presenting token: 2 errors occurred:
	* rpc error: code = Unknown desc = route53: failed to change record set: operation error Route 53: ChangeResourceRecordSets, https response error StatusCode: 400, RequestID: 0d20697b-e5c8-4f8d-ab54-d93c7b20e618, InvalidChangeBatch: [RRSet with DNS name _acme-challenge.dev.some.nice.domain. is not permitted in zone public.domain.]
	* error encountered while presenting token for DNS challenge: rpc error: code = Unknown desc = route53: failed to change record set: operation error Route 53: ChangeResourceRecordSets, https response error StatusCode: 400, RequestID: 0d20697b-e5c8-4f8d-ab54-d93c7b20e618, InvalidChangeBatch: [RRSet with DNS name _acme-challenge.dev.some.nice.domain. is not permitted in zone public.domain.]


	
	
# dig -t CNAME _acme-challenge.dev.some.nice.domain

; <<>> DiG 9.20.0-2ubuntu3-Ubuntu <<>> -t CNAME _acme-challenge.dev.some.nice.domain
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 19814
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;_acme-challenge.dev.some.nice.domain. IN CNAME

;; ANSWER SECTION:
_acme-challenge.dev.some.nice.domain. 60 IN CNAME _acme-challenge.public.domain.

;; Query time: 20 msec
;; SERVER: 10.255.255.254#53(10.255.255.254) (UDP)
;; WHEN: Wed Feb 05 19:47:44 CET 2025
;; MSG SIZE  rcvd: 132
```

and i tried this config:
```
resource "acme_certificate" "private" {
  account_key_pem = acme_registration.reg.account_key_pem
  common_name     = "dev.some.nice.domain"
  subject_alternative_names = [
    "*.dev.some.nice.domain"
  ]
  disable_complete_propagation = true
  recursive_nameservers              = ["ns-1347.awsdns-40.org"] # ns server from some.nice.domain zone
  dns_challenge {
    provider = "route53"
    config = {
      AWS_HOSTED_ZONE_ID = data.aws_route53_zone.domain.zone_id #points to zone with public.domain
      AWS_REGION         = var.aws_region
    }
  }
}
```
so something is missing... 

Example for split-horizon setup works for ages now:
```
resource "acme_certificate" "certificate" {
  account_key_pem = acme_registration.reg.account_key_pem
  common_name     = local.domain
  subject_alternative_names = [
    "*.${local.domain}"
  ]
  disable_complete_propagation = true
  dns_challenge {
    provider = "route53"
    config = {
      AWS_HOSTED_ZONE_ID = data.aws_route53_zone.domain.zone_id # points to public `copy`.
    }
  }
}
```

**Comment by vancluever** [comment]
@yaroslav-nakonechnikov yeah, the error message for sure indicates that the record cannot be added into the zone that it's attempting to be added to. What happened with your original config (the one that you mentioned had been stable until recently)?

**Comment by yaroslav-nakonechnikov** [comment]
it still works. 

today i'll raise question to AWS support, as i don't understand why it is saying that not enough permissions... but without DNS alias - i can request domain without issues.

**Comment by yaroslav-nakonechnikov** [comment]
allright. 

so, i found. 
and yes, it works:
```
resource "acme_certificate" "some_nice_domain" {
  account_key_pem = acme_registration.reg.account_key_pem
  common_name     = "dev.some.nice.domain"
  subject_alternative_names = [
    "*.some.nice.domain"
  ]
  disable_complete_propagation = true
  recursive_nameservers        = ["1.1.1.1:53", "1.0.0.1:53"]
  dns_challenge {
    provider = "route53"
    config = {
      AWS_HOSTED_ZONE_ID = data.aws_route53_zone.domain.zone_id
      AWS_REGION         = var.aws_region
      AWS_DEFAULT_REGION = var.aws_region
    }
  }
}
```

in the public part i've created:
`*.dev.some.nice.domain.  60      IN      CNAME   public.domain.`

and after that i've retrieved certs.


problem was in recursive_nameservers - firewall was blocking request to public dns server, and errors from terraform output and AWS is not very obvious, so that's why i was stuck at some point. 

If there is a way to add some check about accessing `recursive_nameservers` - would be cool, will save time with debugging!


Thanks you, @vancluever!

**Comment by vancluever** [comment]
@yaroslav-nakonechnikov glad you got it worked out in the end!

WRT DNS error logging, lego does do extensive logging that you _should_ be able to see when setting `TF_LOG=DEBUG`. More info here: https://developer.hashicorp.com/terraform/internals/debugging

I did do a cursory check and it does seem like DNS errors are logged in lego ([link](https://github.com/go-acme/lego/blob/c96a165aa916bd2b9181d94402133fae1b78124f/challenge/dns01/nameserver.go#L267-L292)).

Take care!
